### ISSUE TITLE
(www) Split up www/gatsby-node.js

### ISSUE THREAD

**Issue by tesseralis** [gatsbyjs/gatsby#21209]
## Summary

Split up the bulk of the logic in www/gatsby-node.js into separate files based on content: e.g. docs, starters, packages, showcase, etc.

## Motivation

gatsby-node is nearing 1000 lines. Most of its content is split between two methods, `onCreateNode` and `createPages`, both in charge of creating and populating all the pages of the site. It's very easy to get lost in those files. While ideally we would [split up the website into themes](https://github.com/gatsbyjs/gatsby/issues/19105), the process is pretty difficult. As a stop-gap, we can split up the giant gatsby-node file into logical components to ease the transition later.

## Implementation Details

We would create different files storing different types of information:

```
src/
  utils/
    node/
      docs.js
      showcase.js
      starters.js
      creators.js
      packages.js
```

Each file has an `onCreateNode` and a `createPages` function, much as they would if they were plugins:

```
// showcase.js

exports.onCreateNode = ({ node, actions }) => {
  const { createNodeField } = actions
  if (node.internal.type === `StartersYaml`) {
    // handle node info
    createNodeField(...)
  }
}

exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions
  const { data } = await graphql(`
    allSitesYaml {
      nodes {
        fields { slug }
      }
    }
  `)

  data.allSitesYaml.nodes.forEach(node => {
    createPage({
      // node fields
    })
  })
})
```

In gatsby-node, we pull in all these files and run all their methods:

```
// gatsby-node.js

const showcase = require('./src/utils/showcase.js')
const docs = require('./src/utils/docs.js')
// <import other sections>

const sections = [showcase, docs, ...]

exports.onCreateNode = helpers => {
  sections.forEach(section => section.onCreateNode(helpers))
  // handle stray nodes
}

exports.createPages = async helpers => {
  await Promise.all(sections.map(section => section.createPages(helpers)))
  // create other pages not covered
})
```

## Related Issues

stopgap for: #19105

**Comment by dvrylc** [comment]
I'd like to work on this! Pretty new to working on Gatsby so I'll probably need some help along the way 😄

**Comment by tsriram** [comment]
This is exactly what I was thinking when I was looking at the `gatsby-node` today -- this file needs to be split into multiple files 😄 

If no one has started working on it, I'm happy to do this.

**Comment by tesseralis** [comment]
> This is exactly what I was thinking when I was looking at the `gatsby-node` today -- this file needs to be split into multiple files 😄
> 
> If no one has started working on it, I'm happy to do this.

@tsriram go ahead! Its all yours. 

@dvrylc the gatsby-node is pretty complicated and probably not a good first issue for someone new to gatsby. We appreciate your enthusiasm and hope you’ll hang around—try searching for the issues labelled “good first issue”.

**Comment by dvrylc** [comment]
@tesseralis alright sure no worries 😄 It would be great if you could take a look at the draft PR linked so I have an idea of whether I was on the right track (and maybe continue working on it if it's not too far off). Thanks!

**Comment by tesseralis** [comment]
@dvrylc sorry! I didn't see that PR. That actually looks pretty good, so we can keep going with that PR. Sorry for the confusion, @tsriram.

**Comment by tsriram** [comment]
No problem, @tesseralis 😄 

Your draft PR looks great, @dvrylc 👍

**Comment by tesseralis** [comment]
Fixed as of #21217
